vcd/resource_vcd_vm_vgpu_policy.go

```diff
@@ -0,0 +1,517 @@
+	var vgpuProfileList []map[string]interface{}
+	vgpuProfileMap := make(map[string]interface{})
+
+	vgpuProfileMap["id"] = vgpuProfiles[0].Id
+	vgpuProfileMap["count"] = vgpuProfiles[0].Count
+
+	vgpuProfileList = append(vgpuProfileList, vgpuProfileMap)
+	err := d.Set("vgpu_profile", vgpuProfileList)
+	if err != nil {
+		return diag.Errorf("error setting vgpu profile: %s", err)
+	}
+
+	return nil
+}
+
+func setPvdcClusterScope(d *schema.ResourceData, vgpuPolicy *types.VdcComputePolicyV2) diag.Diagnostics {
+	pvdcClusters := make([]interface{}, len(vgpuPolicy.PvdcVgpuClustersMap))
+	for index, cluster := range vgpuPolicy.PvdcVgpuClustersMap {
+		singleScope := make(map[string]interface{})
+		singleScope["provider_vdc_id"] = cluster.Pvdc.ID
+		clusterSet := convertStringsToTypeSet(cluster.Clusters)
+		singleScope["cluster_names"] = clusterSet
+
+		for _, vmGroup := range vgpuPolicy.PvdcNamedVmGroupsMap {
+			if vmGroup.Pvdc.ID == cluster.Pvdc.ID {
+				singleScope["vm_group_id"] = vmGroup.NamedVmGroups[0][0].ID
+			}
+		}
+
+		pvdcClusters[index] = singleScope
+	}
+
+	err := d.Set("provider_vdc_scope", pvdcClusters)
+	if err != nil {
+		return diag.Errorf("error setting provider_vdc_scope: %s", err)
+	}
+
+	return nil
+}
+
+func getVgpuClustersAndVmGroups(d *schema.ResourceData, vcdClient *VCDClient) ([]types.PvdcVgpuClustersMap, []types.PvdcNamedVmGroupsMap, error) {
+	vgpuClusterSet := d.Get("provider_vdc_scope").(*schema.Set)
+	vgpuClusters := make([]types.PvdcVgpuClustersMap, len(vgpuClusterSet.List()))
+	var namedVmGroups []types.PvdcNamedVmGroupsMap
+	for rangeIndex, vgpuCluster := range vgpuClusterSet.List() {
+		vgpuClusterDefinition := vgpuCluster.(map[string]interface{})
+
+		pvdcId := vgpuClusterDefinition["provider_vdc_id"].(string)
+
+		clusterNameSet := vgpuClusterDefinition["cluster_names"].(*schema.Set)
+		clusterNames := convertSchemaSetToSliceOfStrings(clusterNameSet)
+		providerVdc, err := vcdClient.GetProviderVdcById(pvdcId)
+		if err != nil {
+			return nil, nil, err
+		}
+
+		// Helper function so the cluster names field isn't mandatory. If none are provided, VCD chooses one by default
+		if len(clusterNames) == 0 {
+			vgpuProfile, err := vcdClient.GetVgpuProfileById(d.Get("vgpu_profile.0.id").(string))
+			if err != nil {
+				return nil, nil, err
+			}
+
+			queryParams := url.Values{}
+			queryParams.Add("filter", fmt.Sprintf("(vgpuProfileName==%s;name==%s)", vgpuProfile.VgpuProfile.Name, providerVdc.ProviderVdc.Name))
+			vcenters, err := vcdClient.GetAllVCenters(nil)
+			if err != nil {
+				return nil, nil, err
+			}
+			var resourcePools []*govcd.ResourcePool
+			for _, vcenter := range vcenters {
+				rps, err := vcenter.GetAllResourcePools(queryParams)
+				if err != nil {
+					return nil, nil, err
+				}
+				resourcePools = append(resourcePools, rps...)
+			}
+			if len(resourcePools) == 0 {
+				return nil, nil, fmt.Errorf("no resource pools for the provider vdc were found")
```

Comment:
nit
```suggestion
				return nil, nil, fmt.Errorf("no resource pools for the Provider VDC were found")
```

Reply:
Removed this chunk of code so it's irrelevant now.